Microsoft.Identity.Web/Client/TokenCacheProviders/Session/MSALPerUserSessionTokenCacheProvider.cs

```diff
@@ -0,0 +1,233 @@
+﻿/*
+ The MIT License (MIT)
+
+Copyright (c) 2015 Microsoft Corporation
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.Identity.Client;
+using System.Diagnostics;
+using System.Security.Claims;
+using System.Threading;
+
+namespace Microsoft.Identity.Web.Client.TokenCacheProviders
+{
+    /// <summary>
+    /// This is a MSAL's TokenCache implementation for one user. It uses Http session as a backend store
+    /// </summary>
+    public class MSALPerUserSessionTokenCacheProvider : IMSALUserTokenCacheProvider
+    {
+        private static ReaderWriterLockSlim SessionLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+
+        /// <summary>
+        /// Once the user signes in, this will not be null and can be ontained via a call to ClaimsPrincipal.Current
+        /// </summary>
+        internal ClaimsPrincipal SignedInUser;
```

Comment:
why do we keep the ClaimsPrincipal, rather than the MsalObjectId?

Reply:
Because the middleware populates CliamsPrincipal.Current. If this cache is instantiated without a claims principal, it allows us to hook the user token cache to the currently signed in user just like that.  The parameterized constructor is useful for test cases

---
In reply to: [267391104](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/65#discussion_r267391104) [](ancestors = 267391104)